docs/configdoc.md
# L3AFD Config Options Documentation

See [l3afd.cfg](https://github.com/l3af-project/l3afd/blob/main/config/l3afd.cfg) for a full example configuration.


```
[DEFAULT]

[l3afd]
pid-file: ./l3afd.pid
datacenter: dc
bpf-dir: /dev/shm
bpf-log-dir:
shutdown-timeout: 1s
http-client-timeout: 10s
max-ebpf-restart-count: 3
bpf-chaining-enabled: true
swagger-api-enabled: false
# PROD | DEV
environment: PROD
....
```

### Below is the detailed documentation for each field


## [l3afd]

| FieldName     | Default                  | Description                                                                                                                                              | Required        |
| ------------- |--------------------------|----------------------------------------------------------------------------------------------------------------------------------------------------------| --------------- |
|pid-file| `"/var/l3afd/l3afd.pid"` | The path to the l3afd.pid file which contains process id of L3afd                                                                                        | Yes |
|datacenter| `"dc"`                   | Name of Datacenter                                                                                                                                       | Yes |
|bpf-dir| `"/dev/shm"`             | Absolute Path where eBPF packages are to be extracted                                                                                                    | Yes |
|bpf-log-dir| `""`                     | Absolute Path for log files, which is passed to applications on the command line. L3afd does not store any logs itself.                                  | No |
|kernel-major-version| `"5"`                    | Major version of the kernel required to run eBPF programs (Linux Only)                                                                                   | No |
|kernel-minor-version| `"1"`                    | Minor version of the kernel required to run eBPF programs (Linux Only)                                                                                   | No |
|shutdown-timeout| `"1s"`                   | Maximum amount of time allowed for l3afd to gracefully stop. After shutdown-timeout, l3afd will exit even if it could not stop applications.             | No |
|http-client-timeout| `"10s"`                  | Maximum amount of time allowed to get HTTP response headers when fetching a package from a repository                                                    | No |
|max-nf-restart-count| `"3"`                    | Maximum number of tries to restart eBPF applications if they are not running                                                                             | No |
|bpf-chaining-enabled| `"true"`                 | Boolean to set bpf-chaining. For more info about bpf chaining check [L3AF_KFaaS.pdf](https://github.com/l3af-project/l3af-arch/blob/main/L3AF_KFaaS.pdf) | Yes |
|swagger-api-enabled| `"false"`                | Whether the swagger API is enabled or not.  For more info see [swagger.md](https://github.com/l3af-project/l3afd/blob/main/docs/swagger.md)              | No |
|environment| `"PROD"`                 | If set to anything other than "PROD", mTLS security will not be checked                                                                                  | Yes |
|BpfMapDefaultPath| `"/sys/fs/bpf"`          | The base pin path for eBPF maps                                                                                                                          | Yes |
| file-log-location | `"/var/log/l3afd.log"`   | Location of the log file                                                                                                                                 | No |
| file-log-max-size | `"100"`                  | Max size in megabytes for Log file rotation                                                                                                              | No |
| file-log-max-backups | `"20"`                   | Max size in megabytes for Log file rotation                                                                                                              | No |
| file-log-max-age | `"60"`                   | Max number of days to keep Log files                                                                                                                     | No |
| json-format-logs | `"false"`                | Logs to be printed as a JSON object                                                                                                                      | No |

## [ebpf-repo]
| FieldName     | Default                    | Description     | Required |
| ------------- |----------------------------| --------------- |----------|
|url| `"file:///var/l3afd/repo"` |Default repository from which to download eBPF packages| Yes      |

## [web]

| FieldName          | Default       | Description     | Required |
|--------------------| ------------- | --------------- |----------|
| metrics-addr       |`"0.0.0.0:8898"`|Prometheus endpoint for pulling/scraping the metrics.  For more info about Prometheus see [prometheus.io](https://prometheus.io/) | Yes      |
| ebpf-poll-interval |`"30s"`|Periodic interval at which to scrape metrics using Prometheus| No       |
| n-metric-samples   |`"20"`|Number of Metric Samples| No       |


## [xdp-root]
This section is needed when bpf-chaining-enabled is set to true.

| FieldName           | Default                  | Description                                                              | Required        |
|---------------------|--------------------------|--------------------------------------------------------------------------| --------------- |
| package-name        | `"xdp-root"`             | Name of subdirectory in which to extract artifact                        | Yes |
| artifact            | `"l3af_xdp_root.tar.gz"` | Filename of xdp-root package. Only tar.gz and .zip formats are supported | Yes |
| ingress-map-name    | `"xdp_root_array"`       | Ingress map name of xdp-root program                                     | Yes |
| command             | `"xdp_root"`             | Command to run xdp-root program                                          | Yes |
| version             | `"latest"`               | Version of xdp-root program                                              | Yes |
| object-file         | `"xdp_root.bpf.o"`      | File containing the object code for xdp-root program                     | Yes |
| entry-function-name | `"xdp_root"`             | Name of the function that begins the XDP-root program                    | Yes |


## [tc-root]
This section is needed when bpf-chaining-enabled is set to true.

| FieldName                   | Default                    | Description                                                                                                                               | Required        |
|-----------------------------|----------------------------|-------------------------------------------------------------------------------------------------------------------------------------------| --------------- |
| pakage-name                 | `"tc-root"`                | Name of subdirectory in which to extract artifact                                                                                         | Yes |
| artifact                    | `"l3af_tc_root.tar.gz"`    | Filename of tc_root package                                                                                                               | Yes |
| ingress-map-name            | `"tc_ingress_root_array"`  | Ingress map name of tc_root program                                                                                                       | Yes |
| egress-map-name             | `"tc_egress_root_array"`   | Egress map name of tc_root program,for more info about ingress/egress check [cilium](https://docs.cilium.io/en/v1.9/concepts/ebpf/intro/) | Yes |
| command                     | `"tc_root"`                | Command to run tc_root program                                                                                                            | Yes |
| version                     | `"latest"`                 | Version of tc_root program                                                                                                                | Yes |
| ingress-object-file         | `"tc_root_ingress.bpf.o"` | File containing the object code for tc-root ingress program                                                                               | Yes |
| egress-object-file          | `"tc_root_egress.bpf.o"`  | File containing the object code for tc-root egress program                                                                                | Yes |
| ingress-entry-function-name | `"tc_ingress_root"`        | Name of the function that begins the tc-root ingress program                                                                              | Yes |
| egress-entry-function-name  | `"tc_egress_root"`         | Name of the function that begins the tc-root egress program                                                                               | Yes |


## [ebpf-chain-debug]
| FieldName | Default            | Description                                                    | Required |
|-----------|--------------------|----------------------------------------------------------------|----------|
| addr      | `"localhost:8899"` | Hostname and Port of chaining debug REST API                   | No       |
| enabled   | `"false"`          | Boolean to check ebpf chaining debug details is enabled or not | No       |

## [l3af-configs]
| FieldName     | Default       | Description     | Required |
| ------------- | ------------- | --------------- |----------|
|restapi-addr|`"localhost:53000"`| Hostname and Port of l3af-configs REST API | No       |

## [l3af-config-store]
| FieldName     | Default       | Description     | Required        |
| ------------- | ------------- | --------------- | --------------- |
|filename|`"/etc/l3afd/l3af-config.json"`|Absolute path of persistent config file where we are storing L3afBPFPrograms objects. For more info see [models](https://github.com/l3af-project/l3afd/blob/main/models/l3afd.go)| Yes |

## [mtls]
| FieldName     | Default                            | Description                                                                                                                                                                                                                  | Required |
| ------------- |------------------------------------|------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------|----------|
|enabled| `"true"`                           | Boolean controlling whether mTLS is enabled or not on the REST API exposed by l3afd                                                                                                                                          | No       |
|min-tls-version| `"1.3"`                            | Minimum tls version allowed                                                                                                                                                                                                  | No       |
|cert-dir| `"/etc/l3afd/certs"`               | Absolute path of CA certificates. On Linux this points to a filesystem directory, but on Windows it can point to a [certificate store](https://docs.microsoft.com/en-us/windows-hardware/drivers/install/certificate-stores) | No       |
|server-crt-filename| `"server.crt"`                     | Server's ca certificate filename                                                                                                                                                                                             | No       |
|server-key-filename| `"server.key"`                     | Server's mtls key filename                                                                                                                                                                                                   | No       |
|cert-expiry-warning-days| `"30"`                             | How many days before expiry you want warning                                                                                                                                                                                 | No       |
|san-match-rules| `".*l3af.l3af.io,^l3afd.l3af.io$"` | List of domain names (exact match) or regular expressions to validate client SAN DNS Names against                                                                                                                                                                  | No      |


config/config.go
// Copyright Contributors to the L3AF Project.
// SPDX-License-Identifier: Apache-2.0

// Package config provides primitives for l3afd configuration ( i.e. l3afd.cfg) file.
package config

import (
	"crypto/tls"
	"fmt"
	"strings"
	"time"

	"github.com/robfig/config"
	"github.com/rs/zerolog/log"
)

const (
	ENV_PROD = "PROD"
)

type Config struct {
	PIDFilename         string
	DataCenter          string
	BPFDir              string
	BPFLogDir           string
	MinKernelMajorVer   int
	MinKernelMinorVer   int
	EBPFRepoURL         string
	HttpClientTimeout   time.Duration
	MaxEBPFReStartCount int
	Environment         string
	BpfMapDefaultPath   string
	// Flag to enable chaining with root program
	BpfChainingEnabled bool

	FileLogLocation   string
	FileLogMaxSize    int
	FileLogMaxBackups int
	FileLogMaxAge     int
	JSONFormatLogs    bool

	// stats
	// Prometheus endpoint for pull/scrape the metrics.
	MetricsAddr      string
	EBPFPollInterval time.Duration
	NMetricSamples   int

	ShutdownTimeout time.Duration

	SwaggerApiEnabled bool

	// XDP Root program details.
	XDPRootPackageName       string
	XDPRootArtifact          string
	XDPRootMapName           string
	XDPRootCommand           string
	XDPRootVersion           string
	XDPRootObjectFile        string
	XDPRootEntryFunctionName string

	// TC Root program details.
	TCRootPackageName              string
	TCRootArtifact                 string
	TCRootIngressMapName           string
	TCRootEgressMapName            string
	TCRootCommand                  string
	TCRootVersion                  string
	TCRootIngressObjectFile        string
	TCRootEgressObjectFile         string
	TCRootIngressEntryFunctionName string
	TCRootEgressEntryFunctionName  string

	// ebpf chain details
	EBPFChainDebugAddr    string
	EBPFChainDebugEnabled bool

	// l3af configs to listen addrs
	L3afConfigsRestAPIAddr string

	// l3af config store
	L3afConfigStoreFileName string

	// mTLS
	MTLSEnabled               bool
	MTLSMinVersion            uint16
	MTLSCertDir               string
	MTLSCACertFilename        string
	MTLSServerCertFilename    string
	MTLSServerKeyFilename     string
	MTLSCertExpiryWarningDays int
	MTLSSANMatchRules         []string

	// graceful-restart
	TimetoRestart      int
	BasePath           string
	RestartArtifactURL string
	VersionLimit       int
}

// ReadConfig - Initializes configuration from file
func ReadConfig(configPath string) (*Config, error) {

	log.Info().Msgf("Reading configuration from: %s", configPath)
	confReader, configErr := config.ReadDefault(configPath)
	if configErr != nil {
		log.Fatal().Err(configErr).Msgf("Could not open config file %q", configPath)
	}
	minTLSVersion, err := loadTLSVersion(confReader, "min-tls-version")
	if err != nil {
		return nil, err
	}

	return &Config{
		PIDFilename:                    LoadConfigString(confReader, "l3afd", "pid-file"),
		DataCenter:                     LoadConfigString(confReader, "l3afd", "datacenter"),
		BPFDir:                         LoadConfigString(confReader, "l3afd", "bpf-dir"),
		BPFLogDir:                      LoadOptionalConfigString(confReader, "l3afd", "bpf-log-dir", ""),
		MinKernelMajorVer:              LoadOptionalConfigInt(confReader, "l3afd", "kernel-major-version", 5),
		MinKernelMinorVer:              LoadOptionalConfigInt(confReader, "l3afd", "kernel-minor-version", 15),
		FileLogLocation:                LoadOptionalConfigString(confReader, "l3afd", "file-log-location", ""),
		FileLogMaxSize:                 LoadOptionalConfigInt(confReader, "l3afd", "file-log-max-size", 100),
		FileLogMaxBackups:              LoadOptionalConfigInt(confReader, "l3afd", "file-log-max-backups", 20),
		FileLogMaxAge:                  LoadOptionalConfigInt(confReader, "l3afd", "file-log-max-age", 60),
		JSONFormatLogs:                 LoadOptionalConfigBool(confReader, "l3afd", "json-format-logs", false),
		EBPFRepoURL:                    LoadConfigString(confReader, "ebpf-repo", "url"),
		HttpClientTimeout:              LoadOptionalConfigDuration(confReader, "l3afd", "http-client-timeout", 30*time.Second),
		MaxEBPFReStartCount:            LoadOptionalConfigInt(confReader, "l3afd", "max-ebpf-restart-count", 3),
		BpfChainingEnabled:             LoadConfigBool(confReader, "l3afd", "bpf-chaining-enabled"),
		MetricsAddr:                    LoadConfigString(confReader, "web", "metrics-addr"),
		EBPFPollInterval:               LoadOptionalConfigDuration(confReader, "web", "ebpf-poll-interval", 30*time.Second),
		NMetricSamples:                 LoadOptionalConfigInt(confReader, "web", "n-metric-samples", 20),
		ShutdownTimeout:                LoadOptionalConfigDuration(confReader, "l3afd", "shutdown-timeout", 25*time.Second),
		SwaggerApiEnabled:              LoadOptionalConfigBool(confReader, "l3afd", "swagger-api-enabled", false),
		Environment:                    LoadOptionalConfigString(confReader, "l3afd", "environment", ENV_PROD),
		BpfMapDefaultPath:              LoadConfigString(confReader, "l3afd", "BpfMapDefaultPath"),
		XDPRootPackageName:             loadXDPRootPackageName(confReader),
		XDPRootArtifact:                loadXDPRootArtifact(confReader),
		XDPRootMapName:                 loadXDPRootIngressMapName(confReader),
		XDPRootCommand:                 loadXDPRootCommand(confReader),
		XDPRootVersion:                 loadXDPRootVersion(confReader),
		XDPRootObjectFile:              LoadOptionalConfigString(confReader, "xdp-root", "object-file", "xdp_root.bpf.o"),
		XDPRootEntryFunctionName:       LoadOptionalConfigString(confReader, "xdp-root", "entry-function-name", "xdp_root"),
		TCRootPackageName:              loadTCRootPackageName(confReader),
		TCRootArtifact:                 loadTCRootArtifact(confReader),
		TCRootIngressMapName:           loadTCRootIngressMapName(confReader),
		TCRootEgressMapName:            loadTCRootEgressMapName(confReader),
		TCRootCommand:                  loadTCRootCommand(confReader),
		TCRootVersion:                  loadTCRootVersion(confReader),
		TCRootIngressObjectFile:        LoadOptionalConfigString(confReader, "tc-root", "ingress-object-file", "tc_root_ingress.bpf.o"),
		TCRootEgressObjectFile:         LoadOptionalConfigString(confReader, "tc-root", "egress-object-file", "tc_root_egress.bpf.o"),
		TCRootIngressEntryFunctionName: LoadOptionalConfigString(confReader, "tc-root", "ingress-entry-function-name", "tc_ingress_root"),
		TCRootEgressEntryFunctionName:  LoadOptionalConfigString(confReader, "tc-root", "egress-entry-function-name", "tc_egress_root"),
		EBPFChainDebugAddr:             LoadOptionalConfigString(confReader, "ebpf-chain-debug", "addr", "localhost:8899"),
		EBPFChainDebugEnabled:          LoadOptionalConfigBool(confReader, "ebpf-chain-debug", "enabled", false),
		L3afConfigsRestAPIAddr:         LoadOptionalConfigString(confReader, "l3af-configs", "restapi-addr", "localhost:53000"),
		L3afConfigStoreFileName:        LoadConfigString(confReader, "l3af-config-store", "filename"),
		MTLSEnabled:                    LoadOptionalConfigBool(confReader, "mtls", "enabled", true),
		MTLSMinVersion:                 minTLSVersion,
		MTLSCertDir:                    LoadOptionalConfigString(confReader, "mtls", "cert-dir", ""),
		MTLSCACertFilename:             LoadOptionalConfigString(confReader, "mtls", "cacert-filename", "ca.pem"),
		MTLSServerCertFilename:         LoadOptionalConfigString(confReader, "mtls", "server-cert-filename", "server.crt"),
		MTLSServerKeyFilename:          LoadOptionalConfigString(confReader, "mtls", "server-key-filename", "server.key"),
		MTLSCertExpiryWarningDays:      LoadOptionalConfigInt(confReader, "mtls", "cert-expiry-warning-days", 30),
		MTLSSANMatchRules:              strings.Split(LoadOptionalConfigString(confReader, "mtls", "san-match-rules", ""), ","),
		BasePath:                       LoadOptionalConfigString(confReader, "graceful-restart", "basepath", "/usr/local/l3afd"),
		TimetoRestart:                  LoadOptionalConfigInt(confReader, "graceful-restart", "time-to-restart", 7),
		VersionLimit:                   LoadOptionalConfigInt(confReader, "graceful-restart", "version-limit", 100),
		RestartArtifactURL:             LoadOptionalConfigString(confReader, "graceful-restart", "restart-artifacts-url", "file:///srv/l3afd"),
	}, nil
}

func loadTLSVersion(cfgRdr *config.Config, fieldName string) (uint16, error) {
	ver := strings.TrimSpace(LoadOptionalConfigString(cfgRdr, "mTLS", fieldName, "TLS_1.3"))
	switch ver {
	case "", "Default", "default":
		return tls.VersionTLS13, nil
	case "TLS_1.2":
		return tls.VersionTLS12, nil
	case "TLS_1.3":
		return tls.VersionTLS13, nil
	default:
		return 0, fmt.Errorf("unsupported TLS version: %s; use: TLS_1.{2,3}", ver)
	}
}

func loadXDPRootPackageName(cfgRdr *config.Config) string {
	xdpRootPackageName := LoadOptionalConfigString(cfgRdr, "xdp-root-program", "name", "")
	if xdpRootPackageName == "" {
		xdpRootPackageName = LoadOptionalConfigString(cfgRdr, "xdp-root", "package-name", "xdp-root")
	}
	return xdpRootPackageName
}

func loadXDPRootArtifact(cfgRdr *config.Config) string {
	xdpRootArtifactName := LoadOptionalConfigString(cfgRdr, "xdp-root-program", "artifact", "")
	if xdpRootArtifactName == "" {
		xdpRootArtifactName = LoadOptionalConfigString(cfgRdr, "xdp-root", "artifact", "l3af_xdp_root.tar.gz")
	}
	return xdpRootArtifactName
}

func loadXDPRootIngressMapName(cfgRdr *config.Config) string {
	xdpRootIngressMapName := LoadOptionalConfigString(cfgRdr, "xdp-root-program", "ingress-map-name", "")
	if xdpRootIngressMapName == "" {
		xdpRootIngressMapName = LoadOptionalConfigString(cfgRdr, "xdp-root", "ingress-map-name", "xdp_root_array")
	}
	return xdpRootIngressMapName
}

func loadXDPRootCommand(cfgRdr *config.Config) string {
	xdpRootCommand := LoadOptionalConfigString(cfgRdr, "xdp-root-program", "command", "")
	if xdpRootCommand == "" {
		xdpRootCommand = LoadOptionalConfigString(cfgRdr, "xdp-root", "command", "xdp_root")
	}
	return xdpRootCommand
}

func loadXDPRootVersion(cfgRdr *config.Config) string {
	xdpRootVersion := LoadOptionalConfigString(cfgRdr, "xdp-root-program", "version", "")
	if xdpRootVersion == "" {
		xdpRootVersion = LoadOptionalConfigString(cfgRdr, "xdp-root", "version", "latest")
	}
	return xdpRootVersion
}

func loadTCRootPackageName(cfgRdr *config.Config) string {
	tcRootPackageName := LoadOptionalConfigString(cfgRdr, "tc-root-program", "name", "")
	if tcRootPackageName == "" {
		tcRootPackageName = LoadOptionalConfigString(cfgRdr, "tc-root", "package-name", "tc-root")
	}
	return tcRootPackageName
}

func loadTCRootArtifact(cfgRdr *config.Config) string {
	tcRootArtifactName := LoadOptionalConfigString(cfgRdr, "tc-root-program", "artifact", "")
	if tcRootArtifactName == "" {
		tcRootArtifactName = LoadOptionalConfigString(cfgRdr, "tc-root", "artifact", "l3af_tc_root.tar.gz")
	}
	return tcRootArtifactName
}

func loadTCRootIngressMapName(cfgRdr *config.Config) string {
	tcRootIngressMapName := LoadOptionalConfigString(cfgRdr, "tc-root-program", "ingress-map-name", "")
	if tcRootIngressMapName == "" {
		tcRootIngressMapName = LoadOptionalConfigString(cfgRdr, "tc-root", "ingress-map-name", "tc_ingress_root_array")
	}
	return tcRootIngressMapName
}

func loadTCRootEgressMapName(cfgRdr *config.Config) string {
	tcRootEgressMapName := LoadOptionalConfigString(cfgRdr, "tc-root-program", "egress-map-name", "")
	if tcRootEgressMapName == "" {
		tcRootEgressMapName = LoadOptionalConfigString(cfgRdr, "tc-root", "egress-map-name", "tc_egress_root_array")
	}
	return tcRootEgressMapName
}

func loadTCRootCommand(cfgRdr *config.Config) string {
	tcRootCommand := LoadOptionalConfigString(cfgRdr, "tc-root-program", "command", "")
	if tcRootCommand == "" {
		tcRootCommand = LoadOptionalConfigString(cfgRdr, "tc-root", "command", "tc_root")
	}
	return tcRootCommand
}

func loadTCRootVersion(cfgRdr *config.Config) string {
	tcRootVersion := LoadOptionalConfigString(cfgRdr, "tc-root-program", "version", "")
	if tcRootVersion == "" {
		tcRootVersion = LoadOptionalConfigString(cfgRdr, "tc-root", "version", "latest")
	}
	return tcRootVersion
}
